### PR TITLE
fix finalizer_test.go

### PR DESCRIPTION
**Pull request by lesterli**
## Summary

This PR completed the first task in https://github.com/babylonchain/optimism/issues/24 to fix the failed tests in the `finalizer_test.go`.

We updated `finalizer.go` to check the finality on each L2 block in this PR https://github.com/babylonchain/optimism/pull/23#issue-2404494291, which

* added a `l2Fetcher` to fetch each L2 block
   
	```Go
	type FinalizerL2Interface interface {
		L2BlockRefByNumber(context.Context, uint64) (eth.L2BlockRef, error)
	}
	```
* updated the SDK query function to check a range of L2 blocks

	```Go
	type BabylonFinalityClient interface {
		QueryBlockRangeBabylonFinalized(queryBlocks []*cwclient.L2Block) (*uint64, error)
	}
	```

So we added two helper functions in the `finalizer_test.go`.

```Go
func mockL2BlockRefByNumberWithTimes(l2F *testutils.MockL2Client, times int, refs ...eth.L2BlockRef)
func mockQueryBlockRangeBabylonFinalizedWithTimes(sdkClient *mocks.MockISdkClient, times int, refs ...eth.L2BlockRef)
```

Also, we added a function in the `mock_l2.go`  to support the mock times.

```Go
func (m *MockL2Client) ExpectL2BlockRefByNumberWithTimes(num uint64, ref eth.L2BlockRef, times int, err error)
```

## Test Plan

```
make lint-go

cd op-node
make test
```

Also, connect with FP e2e tests locally and make sure all test cases are passed.

### CODE REVIEW

File: op-node/rollup/finality/finalizer.go [hunk 1]
```diff
@@ -366,13 +365,6 @@ func (fi *Finalizer) findLastFinalizedL2BlockWithConsecutiveQuorom(
 	return nil
 }
 
-func (fi *Finalizer) updateFinalized(lastFinalizedBlock eth.L2BlockRef, fdL1Block eth.BlockID) (eth.L2BlockRef, eth.BlockID) {
-	finalizedL2 := lastFinalizedBlock
-	finalizedDerivedFrom := fdL1Block
-	fi.log.Debug("set finalized block", "l2_block", finalizedL2, "derived_from", finalizedDerivedFrom)
```
Comment (bap2pecs) [inline op-node/rollup/finality/finalizer.go:372]: why deleting this log? I think we should move it inside the if condition

File: op-node/rollup/finality/finalizer.go [hunk 2]
```diff
@@ -243,19 +243,18 @@ func (fi *Finalizer) tryFinalize() {
 	finalizedL2 := fi.lastFinalizedL2 // may be zeroed if nothing was finalized since startup.
 	var finalizedDerivedFrom eth.BlockID
 	// go through the latest inclusion data, and find the last L2 block that was derived from a finalized L1 block
+	fi.log.Debug("tryFinalize", "finalityData", fi.finalityData, "finalizedL2", finalizedL2)
 	for _, fd := range fi.finalityData {
 		if fd.L2Block.Number > finalizedL2.Number && fd.L1Block.Number <= fi.finalizedL1.Number {
 			lastFinalizedBlock := fi.findLastFinalizedL2BlockWithConsecutiveQuorom(fd.L2Block.Number, finalizedL2.Number)
+			fi.log.Debug("finalizing L2 block", "l2_block", fd.L2Block, "derived_from", fd.L1Block, "last_finalized_block", lastFinalizedBlock)
```
Comment (bap2pecs) [inline op-node/rollup/finality/finalizer.go]: we can remove this

File: op-node/rollup/finality/finalizer.go [hunk 3]
```diff
@@ -243,19 +243,18 @@ func (fi *Finalizer) tryFinalize() {
...
 
 			// set finalized block(s)
-			if lastFinalizedBlock != nil {
-				finalizedL2, finalizedDerivedFrom = fi.updateFinalized(*lastFinalizedBlock, lastFinalizedBlock.L1Origin)
+			if lastFinalizedBlock != nil && lastFinalizedBlock.Number == fd.L2Block.Number {
```
Comment (bap2pecs) [inline op-node/rollup/finality/finalizer.go]: this change is gonna break the OP e2e test. any idea how to fix it?

let's make sure we test against OP e2e tests whenever we make changes in the OP codebase

File: op-node/rollup/finality/finalizer_test.go [hunk 4]
```diff
@@ -82,6 +85,7 @@ func TestEngineQueue_Finalize(t *testing.T) {
 		L1Origin:       refA.ID(),
 		SequenceNumber: 0,
 	}
+
```
Comment (bap2pecs) [inline op-node/rollup/finality/finalizer_test.go]: let's remove this to minimize unnecessary changes to the OP codebase

File: op-node/rollup/finality/finalizer_test.go [hunk 5]
```diff
@@ -190,8 +201,48 @@ func TestEngineQueue_Finalize(t *testing.T) {
...
+		l2F.ExpectL2BlockRefByNumber(refA1.Number, refA1, nil)
+		l2F.ExpectL2BlockRefByNumber(refB0.Number, refB0, nil)
+		l2F.ExpectL2BlockRefByNumber(refB1.Number, refB1, nil)
+		l2F.ExpectL2BlockRefByNumber(refC0.Number, refC0, nil)
+		l2F.ExpectL2BlockRefByNumber(refC1.Number, refC1, nil)
+
+		ctl := gomock.NewController(t)
+		defer ctl.Finish()
+		sdkClient := mocks.NewMockISdkClient(ctl)
+		queryBlocks := make([]*cwclient.L2Block, refC1.Number)
```
Comment (bap2pecs) [inline op-node/rollup/finality/finalizer_test.go]: could you add more context in the PR description to explain why it failed and what's the fix? it will be helpful to pass down knowledge when someone look at the PR again. will also make the review process easier

for this, could you explain why we need this specific 5 blocks to be mocked?

File: op-node/rollup/finality/finalizer_test.go [hunk 6]
```diff
@@ -224,8 +275,53 @@ func TestEngineQueue_Finalize(t *testing.T) {
 		l1F.ExpectL1BlockRefByNumber(refD.Number, refD, nil) // to check finality signal
 		l1F.ExpectL1BlockRefByNumber(refD.Number, refD, nil) // to check what was derived from (same in this case)
 
+		l2F := &testutils.MockL2Client{}
+		defer l2F.AssertExpectations(t)
+		l2F.ExpectL2BlockRefByNumber(refA1.Number, refA1, nil)
```
Comment (bap2pecs) [inline op-node/rollup/finality/finalizer_test.go]: let's avoid repeating ourselves. you can extract it to be helper function that takes variable length number of ref blocks:

```
l2F = helperFunc(refA1, refB0, .... refC1)
```

File: op-node/rollup/finality/finalizer_test.go [hunk 7]
```diff
@@ -224,8 +275,53 @@ func TestEngineQueue_Finalize(t *testing.T) {
...
+		l2F.ExpectL2BlockRefByNumber(refB0.Number, refB0, nil)
+		l2F.ExpectL2BlockRefByNumber(refB1.Number, refB1, nil)
+		l2F.ExpectL2BlockRefByNumber(refC0.Number, refC0, nil)
+		l2F.ExpectL2BlockRefByNumber(refC1.Number, refC1, nil)
+
+		ctl := gomock.NewController(t)
+		defer ctl.Finish()
+		sdkClient := mocks.NewMockISdkClient(ctl)
+		queryBlocks := make([]*cwclient.L2Block, refC1.Number)
+		queryBlocks[0] = &cwclient.L2Block{
```
Comment (bap2pecs) [inline op-node/rollup/finality/finalizer_test.go]: same here. please create helper function instead of copy&paste

File: op-node/rollup/finality/finalizer_test.go [hunk 8]
```diff
@@ -224,8 +275,53 @@ func TestEngineQueue_Finalize(t *testing.T) {
...
+			BlockHeight:    refB1.Number,
+			BlockHash:      refB1.Hash.String(),
+			BlockTimestamp: refB1.Time,
+		}
+		queryBlocks[3] = &cwclient.L2Block{
+			BlockHeight:    refC0.Number,
+			BlockHash:      refC0.Hash.String(),
+			BlockTimestamp: refC0.Time,
+		}
+		queryBlocks[4] = &cwclient.L2Block{
```
Comment (bap2pecs) [inline op-node/rollup/finality/finalizer_test.go]: also could you explain why we call `ExpectL2BlockRefByNumber` on 10 blocks but still only mock 5 blocks here?

File: op-node/rollup/finality/finalizer.go (hunk 3 above)
Comment (lesterli) [reply op-node/rollup/finality/finalizer.go]: In the OP e2e test, FP can't predict how many l2 blocks will be derived, so we need to revert this change.

File: op-node/rollup/finality/finalizer_test.go (hunk 6 above)
Comment (lesterli) [reply op-node/rollup/finality/finalizer_test.go]: `helperL2BlockRefByNumberWithTimes`

File: op-node/rollup/finality/finalizer_test.go (hunk 5 above)
Comment (lesterli) [reply op-node/rollup/finality/finalizer_test.go]: updated the PR description

File: op-node/rollup/finality/finalizer.go [hunk 9]
```diff
@@ -243,19 +243,23 @@ func (fi *Finalizer) tryFinalize() {
...
 		if fd.L2Block.Number > finalizedL2.Number && fd.L1Block.Number <= fi.finalizedL1.Number {
 			lastFinalizedBlock := fi.findLastFinalizedL2BlockWithConsecutiveQuorom(fd.L2Block.Number, finalizedL2.Number)
 
 			// set finalized block(s)
 			if lastFinalizedBlock != nil {
-				finalizedL2, finalizedDerivedFrom = fi.updateFinalized(*lastFinalizedBlock, lastFinalizedBlock.L1Origin)
+				finalizedL2 = *lastFinalizedBlock
+				finalizedDerivedFrom = fd.L1Block
 			}
+			fi.log.Debug("set finalized block", "finalized_l2", finalizedL2, "finalized_derived_from", finalizedDerivedFrom, "fd_l2_block", fd.L2Block)
```
Comment (bap2pecs) [inline op-node/rollup/finality/finalizer.go]: if you see the original logic, this debug log is only called when finalizedL2 and finalizedDerivedFrom is set.

File: op-node/rollup/finality/finalizer.go [hunk 10]
```diff
@@ -243,19 +243,23 @@ func (fi *Finalizer) tryFinalize() {
...
 
 			// some blocks in the queried range is not BTC finalized, stop iterating to honor consecutive quorom
-			if lastFinalizedBlock == nil || lastFinalizedBlock.Number != fd.L2Block.Number {
+			if lastFinalizedBlock == nil || finalizedL2.Number != fd.L2Block.Number {
```
Comment (bap2pecs) [inline op-node/rollup/finality/finalizer.go]: why change this?

File: op-node/rollup/finality/finalizer_test.go [hunk 11]
```diff
@@ -190,8 +199,18 @@ func TestEngineQueue_Finalize(t *testing.T) {
 		l1F.ExpectL1BlockRefByNumber(refD.Number, refD, nil)
 		l1F.ExpectL1BlockRefByNumber(refD.Number, refD, nil)
 
+		l2F := &testutils.MockL2Client{}
+		defer l2F.AssertExpectations(t)
+		helperL2BlockRefByNumberWithTimes(t, l2F, 1, refA1, refB0, refB1, refC0, refC1)
```
Comment (bap2pecs) [inline op-node/rollup/finality/finalizer_test.go]: i usually don't see people add "helper" to the name of a helper function. let's name it to `mockL2BlockRefByNumberWithTimes`

File: op-node/rollup/finality/finalizer_test.go [hunk 12]
```diff
@@ -190,8 +199,18 @@ func TestEngineQueue_Finalize(t *testing.T) {
...
+
 		emitter := &testutils.MockEmitter{}
-		fi := NewFinalizer(context.Background(), logger, &rollup.Config{}, l1F, emitter)
+		fi := NewFinalizer(context.Background(), logger, &rollup.Config{BabylonConfig: babylonCfg}, l1F, l2F, emitter)
+
+		ctl := gomock.NewController(t)
+		defer ctl.Finish()
+		sdkClient := mocks.NewMockISdkClient(ctl)
+		fi.babylonFinalityClient = sdkClient
+		helperSdkClientQueryBlockRangeBabylonFinalized(t, sdkClient, 1, refA1, refB0, refB1, refC0, refC1)
```
Comment (bap2pecs) [inline op-node/rollup/finality/finalizer_test.go]: same for this. let's think of a better name

File: op-node/rollup/finality/finalizer_test.go [hunk 13]
```diff
@@ -468,3 +532,21 @@ func TestEngineQueue_Finalize(t *testing.T) {
 		emitter.AssertExpectations(t)
 	})
 }
+
+func helperL2BlockRefByNumberWithTimes(t *testing.T, l2F *testutils.MockL2Client, times int, refs ...eth.L2BlockRef) {
```
Comment (bap2pecs) [inline op-node/rollup/finality/finalizer_test.go]: is `t` used anyway? same for `helperSdkClientQueryBlockRangeBabylonFinalized`

File: op-node/rollup/finality/finalizer.go (hunk 10 above)
Comment (lesterli) [reply op-node/rollup/finality/finalizer.go]: sorry, I made a mistake.

File: op-node/rollup/finality/finalizer_test.go (hunk 13 above)
Comment (lesterli) [reply op-node/rollup/finality/finalizer_test.go]: no, it should be removed

File: op-node/rollup/finality/finalizer_test.go (hunk 12 above)
Comment (lesterli) [reply op-node/rollup/finality/finalizer_test.go]: name it `mockQueryBlockRangeBabylonFinalizedWithTimes`?

File: op-node/rollup/finality/finalizer_test.go (hunk 11 above)
Comment (lesterli) [reply op-node/rollup/finality/finalizer_test.go]: ok

File: op-node/rollup/finality/finalizer.go (hunk 9 above)
Comment (lesterli) [reply op-node/rollup/finality/finalizer.go]: yes, changed.

Comment (bap2pecs) [reply op-node/rollup/finality/finalizer.go]: no my point was "set finalized block" should be printed **only when** finalizedL2 and finalizedDerivedFrom is set

see the original code:
image

but after your changes, it will be printed no matter whether they are changed or not

**Review by bap2pecs — APPROVED** [review]
Elon Musk Travel - ZOilRom6v644PCsqVR
